### PR TITLE
feat(dismissProspect): [BACK-1623] add event rule and sns topic for dismiss prospect event

### PR DESCRIPTION
**Pull request by Herraj**
## Goal
Set up the event rule for `dismiss-prospect` event. 

We're setting up a SNS topic for this event rule to which the Event Bridge will publish to.

[Dev deployment of the rule](https://us-east-1.console.aws.amazon.com/events/home?region=us-east-1#/schemas)

JIRA ticket:
* [BACK-1623]

[BACK-1623]: https://getpocket.atlassian.net/browse/BACK-1623?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .aws/src/event-rules/prospect-events/eventConfig.ts [hunk 1]
```diff
@@ -1,4 +1,8 @@
 export const eventConfig = {
   source: 'prospect-events',
   detailType: ['prospect-generation'],
+  dismissProspect: {
+    source: 'prospect-events',
+    detailType: ['dismiss-prospect'],
+  },
```
Comment (jpetto) [inline .aws/src/event-rules/prospect-events/eventConfig.ts:7]: i'm a little confused on the organization here - should we refactor the `prospect-generation` event to be formatted like the `dismiss-prospect` event?

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts [hunk 2]
```diff
@@ -39,17 +41,33 @@ export class ProspectEvents extends Resource {
...
+      }
+    );
 
-    this.sqsDlq = new sqs.DataAwsSqsQueue(this, `prospect-${config.environment}-dlq`, {
-      name: `ProspectAPI-${config.environment}-Sqs-Translation-Queue-Deadletter`
+    this.snsTopic = new sns.SnsTopic(this, 'prospect-event-topic', {
+      name: `${config.prefix}-ProspectEventTopic`,
     });
 
     this.createProspectEventRules();
```
Comment (jpetto) [inline .aws/src/event-rules/prospect-events/prospectEventRules.ts:64]: similar question here - should this function be refactored to something like `createProspectGenerationEventRules`? should we move the above sqs code into a separate function?

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts [hunk 3]
```diff
@@ -39,17 +41,33 @@ export class ProspectEvents extends Resource {
...
+      `prospect-${config.environment}-dlq`,
+      {
+        name: `ProspectAPI-${config.environment}-Sqs-Translation-Queue-Deadletter`,
+      }
+    );
 
-    this.sqsDlq = new sqs.DataAwsSqsQueue(this, `prospect-${config.environment}-dlq`, {
-      name: `ProspectAPI-${config.environment}-Sqs-Translation-Queue-Deadletter`
+    this.snsTopic = new sns.SnsTopic(this, 'prospect-event-topic', {
+      name: `${config.prefix}-ProspectEventTopic`,
```
Comment (jpetto) [inline .aws/src/event-rules/prospect-events/prospectEventRules.ts:61]: is this sns topic specific to dismissing prospects? or a general sns topic to be used for many/all prospect events?

**Review by jpetto — CHANGES_REQUESTED** [review]
code looks perfectly functional, but i'm wondering about consistency/organization. i also realize i'm asking some general event bridge questions here (e.g. the sns topic).

it seems like maybe the prospect generation event is a special case, whereas dismiss (and potentially future) events will be more standard. is that right?

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts (hunk 3 above)
Comment (sri-kirsh) [reply .aws/src/event-rules/prospect-events/prospectEventRules.ts:61]: this SNS topic is target for this event-rule so we can connect more than 5 targets to this rule if needed (event-bridge restricts 5 targets per rule, so we are adopting this pattern)

right now, this has source as `prospect-events` and detail-type as `dismiss-prospects`
but in future, say we have another event (e.g generate-prospect), we can add them to detail-type. 

we already have a prospect-events rule but we didn't add this target to the existing rule coz it already forwards to another translation SQS/dev event bridge. didn't want to have any underlying implications on that workflow. this wa

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts (hunk 2 above)
Comment (Herraj) [reply .aws/src/event-rules/prospect-events/prospectEventRules.ts:64]: We could do a slightly bigger refactor than that and pull out the SQS creation (for prospect-generation events) and the make the SNS creation a general function. That way the SNS creation is not tied to our dimiss-prospect event.

Comment (jpetto) [reply .aws/src/event-rules/prospect-events/prospectEventRules.ts:64]: i like the direction proposed. i'll leave it up to you to decide on where the line is between a refactor now and future improvements.

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts [hunk 4]
```diff
@@ -39,33 +50,82 @@ export class ProspectEvents extends Resource {
...
-    this.createProspectEventRules();
+    this.createProspectGenerationEventRule();
     this.createPolicyForEventBridgeToSqs();
 
+    // setting up prospect-dismiss event rule
+    this.createProspectEventRule(
+      'Dismiss-Prospect-Events-Rule',
+      eventConfig.prospectDismiss.source,
+      eventConfig.prospectDismiss.detailType
+    );
```
Comment (Herraj) [inline .aws/src/event-rules/prospect-events/prospectEventRules.ts:75]: Should we rename this event to `Prospect-Dimiss-Events-Rule` (being set below in the function on **line 184**) ? 

This is not consistent with the other event rule name on **line 144** below

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts [hunk 5]
```diff
@@ -94,10 +154,83 @@ export class ProspectEvents extends Resource {
...
+  private createProspectEventRule(
+    name: string,
+    source: string,
+    detailType: string[]
+  ) {
+    const targets: PocketEventBridgeTargets[] = [
+      {
+        arn: this.snsTopic.arn,
+        deadLetterArn: this.sqsDlq.arn, //using the same DLQ for all prospect events (prospect-generation and prospect-dismiss as of now)
+        targetId: `${config.prefix}-Prospect-Event-SNS-Target`,
```
Comment (Herraj) [inline .aws/src/event-rules/prospect-events/prospectEventRules.ts:177]: @sri-kirsh I forgot but do we need a unique `targetId` here for each Prospect event that will use the SNS or can we keep it the same for all of them?

Comment (sri-kirsh) [reply .aws/src/event-rules/prospect-events/prospectEventRules.ts:177]: in future, if you have more `prospect-events` that you want to send to this SNS, you can just add them to `detail-type` of this rule in the config file. 

and yes, targetId need to be unique and tied to the `target`. https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_Target.html. multiple rules should be able to use the same target

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts (hunk 4 above)
Comment (jpetto) [reply .aws/src/event-rules/prospect-events/prospectEventRules.ts:75]: i like the `{Entity}-{Action}-Events-Rule` pattern, and would vote to change the name on line 144 if it won't cause any issues.

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts [hunk 6]
```diff
@@ -41,42 +50,75 @@ export class ProspectEvents extends Resource {
...
-    this.sqsDlq = new sqs.DataAwsSqsQueue(
-      this,
-      `prospect-${config.environment}-dlq`,
-      {
-        name: `ProspectAPI-${config.environment}-Sqs-Translation-Queue-Deadletter`,
-      }
+    // create a dlq for all Prospect events
+    this.sqsDlq = this.createSqsForProspectEvents(
+      this.dlqIdForProspectEvents,
+      this.dlqNameForProspectEvents
```
Comment (jpetto) [inline .aws/src/event-rules/prospect-events/prospectEventRules.ts:61]: doesn't need to block as i don't want to bike shed this forever, but maybe moving these special case calls into the `createprospectGenerationEventRule` function?

sub nit-pick: `retrieveSqsForProspectEvents` may be a better name, as this service is _not_ creating those sqs resources (they are created/owned by prospect api - which is confusing and non-standard 😭 ).

File: .aws/src/event-rules/prospect-events/prospectEventRules.ts (hunk 5 above)
Comment (jpetto) [reply .aws/src/event-rules/prospect-events/prospectEventRules.ts:177]: i'm also curious about this. guessing the answer is "no", but 🤷 .

**Review by jpetto — APPROVED** [review]
a few comments that don't have to block. i think this is a good improvement!

**Review by sri-kirsh — APPROVED** [review]

**Pull request merged**